lib/client-api/src/story_store.ts

```diff
@@ -396,18 +406,43 @@ export default class StoryStore {
 
     const storyParametersWithArgTypes = { ...storyParameters, argTypes, __isArgsStory };
 
-    const storyFn: LegacyStoryFn = (runtimeContext: StoryContext) =>
-      getDecorated()({
+    const storyFn: LegacyStoryFn = deprecate(
+      (runtimeContext: StoryContext) =>
+        getDecorated()({
+          ...identification,
+          ...runtimeContext,
+          // Calculate "combined" parameters at render time (NOTE: for perf we could just use combinedParameters from above?)
+          parameters: this.combineStoryParameters(storyParametersWithArgTypes, kind),
+          hooks,
+          args: _stories[id].args,
+          argTypes,
+          globals: this._globals,
+          viewMode: this._selection?.viewMode,
+        }),
+      dedent`
+        \`storyFn\` is deprecated and will be removed in Storybook 7.0.
+
+        https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-storyfn
+      `
+    );
+
+    const unboundStoryFn: LegacyStoryFn = (context: StoryContext) => getDecorated()(context);
+
+    const applyLoaders = async () => {
+      const context = {
         ...identification,
-        ...runtimeContext,
         // Calculate "combined" parameters at render time (NOTE: for perf we could just use combinedParameters from above?)
         parameters: this.combineStoryParameters(storyParametersWithArgTypes, kind),
         hooks,
         args: _stories[id].args,
         argTypes,
         globals: this._globals,
         viewMode: this._selection?.viewMode,
-      });
+      };
+      const loadResults = await Promise.all(loaders.map((loader) => loader(context)));
+      const loaded = Object.assign({}, ...loadResults);
+      return combineParameters(context, { loaded }) as StoryContext;
```

Comment:
Why `combineParameters` here? Would have thought a simple `{ ...context, loaded }` would be fine.